### PR TITLE
more robust error propagation and handling

### LINKED ISSUE #280: Service worker error propogation
Related to https://github.com/penumbra-zone/web/pull/255

As the service worker handles an incoming grpc request, if in handling it raises, this error does not propagate back to the web app.

Easiest way to test is to pull down https://github.com/penumbra-zone/web/pull/279, and decline a tx approval request. It throws in the service worker but the client hangs indefinitely.

### PR DESCRIPTION
**Pull request by turbocrime**
fixes #280 #287 and should make debugging any failures a bit nicer

### CODE REVIEW

**Comment by turbocrime** [comment]
fixes #293

File: packages/transport/src/chrome-runtime/background-connection-manager.ts [hunk 1]
```diff
@@ -72,105 +77,163 @@ export class BackgroundConnectionManager {
...
+        // TODO: revisit this
+        /*
+        setTimeout(() => {
+          if (chrome.runtime.onConnect.hasListener(subListener)) {
+            // handler still hasn't activated, or it would be gone
+            chrome.runtime.onConnect.removeListener(subListener);
+            acont.abort('Timed out subchannel activation');
+          }
+        }, 1000);
+        */
```
Comment (grod220) [inline packages/transport/src/chrome-runtime/background-connection-manager.ts:176]: Can we either delete this or give more info in the TODO as to what this is about and why it needs revisiting?

File: packages/transport/src/chrome-runtime/background-connection-manager.ts [hunk 2]
```diff
@@ -72,105 +77,163 @@ export class BackgroundConnectionManager {
...
+          void stream
             .pipeTo(new WritableStream(new ChromeRuntimeStreamSink(subPort)), {
               signal: acont.signal,
             })
-            .catch(e => {
-              if (e !== subName) throw e;
+            .catch((error: unknown) => {
+              if (error !== channel) {
+                // TODO: handling errors in this promise chain is problematic.
+                // we're in a callback, and can't reliably answer the requester.
```
Comment (grod220) [inline packages/transport/src/chrome-runtime/background-connection-manager.ts:150]: But isn't the postMessage below answering the requestor? When you say it's not reliable, does that mean it only sends sometimes?

File: packages/transport/src/chrome-runtime/background-connection-manager.ts [hunk 3]
```diff
@@ -72,105 +77,163 @@ export class BackgroundConnectionManager {
...
+
+        const { requestId, message: request } = transportRequest;
+
+        /*
+        const serviceTimeout = new Promise<void>((_, reject) =>
+          setTimeout(reject, 1000, 'Timed out service request'),
+        );
+
+        const transportResponse = Promise.race([serviceEntry(request), serviceTimeout])
+        */
```
Comment (grod220) [inline packages/transport/src/chrome-runtime/background-connection-manager.ts:198]: Delete?

File: packages/transport/src/chrome-runtime/background-connection-manager.ts [hunk 4]
```diff
@@ -72,105 +77,163 @@ export class BackgroundConnectionManager {
...
+                error: errorToJson(ConnectError.from(error), undefined),
+              }) as TransportError,
+          );
+
+        transportResponse
+          .then(extResponse => transPort.postMessage(extResponse))
+          .catch(e => {
+            // TODO: abort requests
+            if (!isDisconnectedPortError(e)) throw e;
+          });
```
Comment (grod220) [inline packages/transport/src/chrome-runtime/background-connection-manager.ts:219]: These can be combined right? Something like:

```typescript
serviceEntry(request)
          .then(response => {
            const extResponse =
              response instanceof ReadableStream
                ? initSubChannel({ requestId, stream: response })
                : ({ requestId, message: response } as TransportMessage);
            transPort.postMessage(extResponse);
          })
          .catch((error: unknown) => {
            const tErr = { requestId, error: String(error) } as TransportError;
            if (!isDisconnectedPortError(tErr)) throw error;
          });
```

File: packages/transport/src/chrome-runtime/client-connection-manager.ts [hunk 5]
```diff
@@ -129,20 +128,25 @@ export class ClientConnectionManager {
...
+            const sourcePort = chrome.runtime.connect({ includeTlsChannelId: true, name: subName });
+            const stream = new ReadableStream(new ChromeRuntimeStreamSource(sourcePort));
+            clientPort.postMessage({ requestId, stream }, [stream]);
+          } else throw Error('Unimplemented response kind');
+        } else if (isTransportError(ev)) clientPort.postMessage(ev);
+        else throw Error('Unknown transport from service');
+      } catch (error) {
+        console.error('Error in service listener', error, ev);
+        clientPort.postMessage({ error: String(error) });
+      }
```
Comment (grod220) [inline packages/transport/src/chrome-runtime/client-connection-manager.ts:149]: This multi-conditional block is a bit hard to read. Not sure how to improve though 🤔

File: packages/transport/src/create.ts [hunk 6]
```diff
@@ -45,60 +53,99 @@ const makeAnyServiceImpl: CreateAnyServiceImpl = <S extends ServiceType>(
...
+      callbackError ??= ConnectError.from(error);
+      throw callbackError;
     }
   });
 
-  const request = (msg: AnyMessage) => {
+  const request = async (msg: AnyMessage) => {
+    if (callbackError) throw callbackError;
+    await providerAck;
+    // TODO: timeout ack? webapp needs to handle error states
```
Comment (grod220) [inline packages/transport/src/create.ts:101]: I think a timeout would be helpful. Can be done later I supppose.

File: packages/transport/src/types/channel-messages.ts [hunk 7]
```diff
@@ -1,78 +1,77 @@
...
   connected: boolean;
+  reason?: JsonValue;
 }
 
-export interface TransportMessage extends TransportEvent {
+export interface TransportError extends Partial<TransportEvent> {
+  error: JsonValue;
+}
+
+// transport actual
```
Comment (grod220) [inline packages/transport/src/types/channel-messages.ts:14]: What's this mean?

File: packages/transport/src/types/channel-messages.ts [hunk 8]
```diff
@@ -1,78 +1,77 @@
...
+
+export type TransportData = TransportMessage | TransportStream | TransportInitChannel;
+
+export interface TransportEvent {
   requestId: ReturnType<typeof crypto.randomUUID>;
+  responseOrigin?: string; // currently unused
```
Comment (grod220) [inline packages/transport/src/types/channel-messages.ts:20]: Should we delete then?

File: packages/transport/src/chrome-runtime/background-connection-manager.ts [hunk 9]
```diff
@@ -5,21 +5,33 @@ import {
...
-  tlsChannelId: string;
+  documentId: string | undefined;
+  tlsChannelId: string | undefined;
 }
 
+type UnconditionalServiceAccessFn = (
+  x: JsonValue,
+) => Promise<JsonValue | ReadableStream<JsonValue>>;
+
+type UnconditionalServiceAccess = Record<string, UnconditionalServiceAccessFn>;
```
[the remark this answers is not in the document]
Comment (grod220) [reply packages/transport/src/chrome-runtime/background-connection-manager.ts:36]: Sorry, was this addressed? Or it's just a temporary todo to fix later?

File: packages/transport/src/create.ts [hunk 10]
```diff
@@ -45,60 +54,108 @@ const makeAnyServiceImpl: CreateAnyServiceImpl = <S extends ServiceType>(
...
+  let callbackError: Error | undefined;
+
+  let connected: boolean | undefined;
+
+  let connectionState: (a: boolean, r?: JsonValue) => void;
+  const providerAck = Promise.race([
+    new Promise<void>((_, reject) =>
+      setTimeout(
+        reject,
+        1000,
```
Comment (grod220) [inline packages/transport/src/create.ts:72]: Can we make this an constant at the top of the file (e.g. `const REQUEST_TIMEOUT = 1000`)? or perhaps a part of a config that gets passed in?

File: packages/transport/src/types/channel-messages.ts [hunk 11]
```diff
@@ -1,78 +1,76 @@
...
-  channed: unknown,
-): channed is StreamChannelChunk | StreamChannelEnd =>
-  typeof channed === 'object' &&
-  channed !== null &&
-  'sequence' in channed &&
-  typeof channed.sequence === 'number' &&
-  ('value' in channed || 'done' in channed);
+// guards
+
+const obj = (o: unknown): o is object => typeof o === 'object' && o !== null;
```
Comment (grod220) [inline packages/transport/src/types/channel-messages.ts:50]: maybe `isObj()`?

**Review by grod220 — APPROVED** [review]
Just had a few more comments.

Also, in general, not all of my comments need responses, but some it would be valuable to reply to.

File: packages/transport/src/chrome-runtime/background-connection-manager.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/transport/src/chrome-runtime/background-connection-manager.ts:36]: ah, sorry, i thought i had replied to this. this is todo.

**Comment by turbocrime** [comment]
ugh i was using the vscode ui to reply to things and it looks like it dropped all my comments? so that's why a bunch of things were resolved with no obvious interaction

**Pull request merged**